### PR TITLE
Adding new request for updating transactions in tx-pool

### PR DESCRIPTION
**Pull request by Mikelle**
Closes https://github.com/hermeznetwork/hermez-node/issues/1035

### What does this PR does?

Fixing bug with updating txs in the pool
### How to test?
Create a transaction and then update it.

Before fix:
Screen Shot 2021-07-28 at 3 39 06 PM
After fix:
Screen Shot 2021-07-28 at 3 39 38 PM

There is a file from the integration tests run:
[report.txt](https://github.com/hermeznetwork/hermez-node/files/6937986/report.txt)

## Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Include tests
- [x] Respect code style and lint
- [ ] Update swagger file (if needed)
- [ ] Update documentation (*.md) (if needed)

### CODE REVIEW

File: api/txspool_test.go [hunk 1]
```diff
@@ -193,14 +193,6 @@ func TestPoolTxs(t *testing.T) {
...
 	err = doBadReq("POST", endpoint, jsonTxReader, 400)
 	require.NoError(t, err)
-	// Wrong rq
-	badTx = tc.poolTxsToSend[0]
-	badTx.RqFromIdx = 30
-	jsonTxBytes, err = json.Marshal(badTx)
-	require.NoError(t, err)
-	jsonTxReader = bytes.NewReader(jsonTxBytes)
-	err = doBadReq("POST", endpoint, jsonTxReader, 409)
-	require.NoError(t, err)
```
Comment (Mikelle) [inline api/txspool_test.go:203]: Deleted this test, bcs internally we didn't check the correctness of the RqFromIdx. This test worked before changes, bcs 409 is the code for ItemAlreadyExist. Now we can update txs in the pool, so this test no longer working.

**Review by arnaubennassar — DISMISSED** [review]
Approving although it would be nice to add tests:

- Tx can be updated from `POST /transactions-pool` (change `ToEthAddr` or `ToBJJ` from one of the inserted txs)
- Tx can NOT be updated from `POST /atomic-pool` (change `ToEthAddr` or `ToBJJ` from one of the inserted txs)

File: db/l2db/l2db.go [hunk 2]
```diff
@@ -267,20 +267,22 @@ func (l2db *L2DB) addTxs(txs []common.PoolL2Tx, checkPoolIsFull bool) error {
 	}
 	// Query begins with the insert statement
 	query := queryInsertPart
+	const onConflictQuery = ` ON CONFLICT ON CONSTRAINT tx_id_unique DO UPDATE SET
```
Comment (jeffprestes) [inline db/l2db/l2db.go:270]: Could you explain better how do you update the transaction item in tx-pool and what are the 'updateful' fields?

File: db/l2db/l2db.go [hunk 3]
```diff
@@ -267,20 +267,22 @@ func (l2db *L2DB) addTxs(txs []common.PoolL2Tx, checkPoolIsFull bool) error {
...
+			from_idx = excluded.from_idx, to_idx = excluded.to_idx, to_eth_addr = excluded.to_eth_addr, to_bjj = excluded.to_bjj,
+			token_id = excluded.token_id, amount = excluded.amount, fee = excluded.fee, nonce = excluded.nonce, state = excluded.state,
+			info = excluded.info, signature = excluded.signature, rq_from_idx = excluded.rq_from_idx, rq_to_idx = excluded.rq_to_idx,
+			rq_to_bjj = excluded.rq_to_bjj, rq_token_id = excluded.rq_token_id, rq_amount = excluded.rq_amount, rq_fee = excluded.rq_fee, rq_nonce = excluded.rq_nonce,
+			tx_type = excluded.tx_type, amount_f = excluded.amount_f, client_ip = excluded.client_ip,
+			rq_offset = excluded.rq_offset, max_num_batch = excluded.max_num_batch WHERE tx_pool.atomic_group_id IS NULL`
+
 	if checkPoolIsFull {
```
Comment (jeffprestes) [inline db/l2db/l2db.go:278]: As this is very critical to the hermez-node I agree with @arnaubennassar and I would recommend to add tests to this item

**Review by jeffprestes — CHANGES_REQUESTED** [review]

File: db/l2db/l2db.go [hunk 4]
```diff
@@ -296,6 +291,70 @@ func (l2db *L2DB) addTxs(txs []common.PoolL2Tx, checkPoolIsFull bool) error {
...
+		queryUpdate += "to_eth_addr = ?, "
+		queryVars = append(queryVars, tx.ToEthAddr)
+	}
+
+	if tx.ToBJJ != common.EmptyBJJComp {
+		queryUpdate += "to_bjj = ?, "
+		queryVars = append(queryVars, tx.ToBJJ)
+	}
+
+	if tx.MaxNumBatch != 0 {
```
Comment (arnaubennassar) [inline db/l2db/l2db.go:327]: Updating to "empty" values is legit, so I'd suggest to just set the values the user is sending, just taking in consideration the existing behaviour for inserts, for instance, if I remember well, an empty Ethereum address (0x000...00) is stored as NULL in the DB

**Review by arnaubennassar — CHANGES_REQUESTED** [review]

File: db/l2db/l2db.go (hunk 2 above)
Comment (Mikelle) [reply db/l2db/l2db.go]: Fields, that can be updated are toAccountIndex, toHezEthereumAddress, toBjj, maxNumBatch, signature, type. Other fields are for atomic txs or take part in transaction id generation. Also, tx state cannot be updated from PUT request

File: api/swagger.yml [hunk 5]
```diff
@@ -32,9 +32,10 @@ info:
 
     ### Signatures
 
-    There are two endpoints that require signatures:
+    There are three endpoints that require signatures:
     * POST /account-creation-authorization: signed using an Ethereum key.
     * POST /transactions-pool: signed using BJJ key.
+    * PUT /transactions-pool: signed using BJJ key.
```
Comment (jeffprestes) [inline api/swagger.yml:38]: A question: when the transaction data changes the transaction id changes too as it's based on transaction data hash. Am I correct?

Comment (Mikelle) [reply api/swagger.yml:38]: Yes, you are right. Fields, that take part in the transaction id creation: fromIdx, tokenID, Amount, Nonce, Fee

Comment (jeffprestes) [reply api/swagger.yml:38]: Thanks for clarifying it.

**Review by jeffprestes — APPROVED** [review]

File: api/txspool.go [hunk 6]
```diff
@@ -64,6 +64,72 @@ func (a *API) postPoolTx(c *gin.Context) {
...
+	if isAtomic(receivedTx) {
+		retBadReq(&apiError{
+			Err:  errors.New(ErrNotAtomicTxsInPostPoolTx),
+			Code: ErrNotAtomicTxsInPostPoolTxCode,
+			Type: ErrNotAtomicTxsInPostPoolTxType,
+		}, c)
+		return
+	}
+	if err := a.verifyPoolL2Tx(receivedTx); err != nil {
+		retBadReq(err, c)
```
Comment (ARR552) [inline api/txspool.go:98]: Why this error doesn't use the apiError structure?

**Review by ARR552 — DISMISSED** [review]

File: api/txspool.go (hunk 6 above)
Comment (Mikelle) [reply api/txspool.go:98]: bcs internally in verifyPoolL2Tx function we already have apiError structure

File: api/atomicpool_test.go [hunk 7]
```diff
@@ -153,6 +153,22 @@ func TestAtomicPool(t *testing.T) {
...
 
+	// test that we can't update atomic tx
+	// this test is checking, that this request will return error
+	// bcs of bad signature. Bad signature returned, bcs Rq* fields
+	// are part of the signature, but they are not part of json, which was sent
+	// we need to keep this test in case Rq* fields will be part of the PoolL2Tx json
+	txRepeated1 := atomicGroup.Txs[1]
+	jsonTxBytes, err = json.Marshal(txRepeated1)
+	require.NoError(t, err)
+	fmt.Println(string(jsonTxBytes))
```
[the remark this answers is not in the document]
Comment (Mikelle) [reply api/atomicpool_test.go:164]: Thank you for the notice!

**Review by marioidival — APPROVED** [review]

**Review by arnaubennassar — APPROVED** [review]

**Pull request merged**